nbodykit/dataset.py

```diff
@@ -0,0 +1,849 @@
+    *   See https://gist.github.com/derricw/95eab740e1b08b78c03f
+    
+    Example
+    -------
+    >>> m = numpy.arange(0,100,1).reshape((10,10))
+    >>> n = bin_ndarray(m, new_shape=(5,5), operation=numpy.sum)
+    >>> print(n)
+    [[ 22  30  38  46  54]
+     [102 110 118 126 134]
+     [182 190 198 206 214]
+     [262 270 278 286 294]
+     [342 350 358 366 374]]
+    """
+    if ndarray.shape == new_shape:
+        raise ValueError("why are we re-binning if the new shape equals the old shape?")
+    if ndarray.ndim != len(new_shape):
+        raise ValueError("Shape mismatch: {} -> {}".format(ndarray.shape, new_shape))
+    if numpy.sometrue(numpy.mod(ndarray.shape, new_shape)):
+        args = (str(new_shape), str(ndarray.shape))
+        msg = "desired shape of %s must be integer factor smaller than the old shape %s" %args
+        raise ValueError(msg)
+        
+    compression_pairs = [(d, c//d) for d, c in zip(new_shape, ndarray.shape)]
+    flattened = [l for p in compression_pairs for l in p]
+    ndarray = ndarray.reshape(flattened)
+    if weights is not None: weights = weights.reshape(flattened)
+    
+    for i in range(len(new_shape)):
+        if weights is not None:
+            ndarray = operation(ndarray*weights, axis=-1*(i+1))
+            weights = numpy.sum(weights, axis=-1*(i+1))
+            ndarray /= weights
+        else:
+            ndarray = operation(ndarray, axis=-1*(i+1))
+    return ndarray
+    
+    
+class DataSet(object):
+    """
+    Lightweight class to hold variables at fixed coordinates, i.e.,
+    a grid of (r,mu) or (k,mu) bins 
+    
+    It is modeled after the syntax of ``xray.DataSet``, and is designed 
+    to hold correlation function or power spectrum results (in 1D or 2D)
+    
+    Notes
+    -----
+    *   note that the suffix `_cen` will be appended to the dimensions
+        passed to the constructor, since the ``coords`` arrray holds
+        the bin centers, as constructed from the bin edges
+    
+    
+    Attributes
+    ----------
+    data : array_like
+        a structured array holding the data variables on the coordinate grid
+        
+    mask : array_like
+        a boolean array where `True` elements indicate that that coordinate
+        grid point has a data variable that is `inf` or `NaN`
+        
+    attrs : collections.OrderedDict
+        an ordered dictionary holding any meta data that has been attached
+        to the instance
+        
+    dims : list
+        a list of strings specifying the names of each axis of the input data
+    
+    edges : dict
+         a dict holding the bin edges for each dimension in `dims`
+        
+    coords : dict
+        a dict holding the bin centers for each dimension in `dims`
+    
+    force_index_match : bool
+        if `True`, when indexing using coordinate values, return
+        results for the nearest bin to the value specified
+        
+    sum_only : list of str
```

Comment:
Is there a more descriptive name than sum_only? It is a list of fields do not need to be averaged. At least `sum_only_fields` would be less suggestive of an action.

A possible extension to this list is a dictionary, fieldname: attribute. and attribute can be SUM_ONLY. It requires more typing though.